File: coding-interview-linked-lists-plus/linkedList.ts
```typescript
class Node<T> {
  data: T;
  next: Node<T> | null;

  constructor(data: T) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList<T> {
  head: Node<T> | null;

  constructor() {
    this.head = null;
  }

  // ... (other LinkedList methods like append, prepend etc.  - see below) ...

  reverse(): LinkedList<T> { //Problem 1: Reverse Linked List (Iterative Approach)
    let prev: Node<T> | null = null;
    let curr: Node<T> | null = this.head;
    while (curr) {
      let nextTemp = curr.next;
      curr.next = prev;
      prev = curr;
      curr = nextTemp;
    }
    this.head = prev;
    return this;
  }

    detectCycle(): boolean { //Problem 2: Detect Cycle (Floyd's Tortoise and Hare)
        let slow = this.head;
        let fast = this.head;
    
        while (fast && fast.next) {
          slow = slow!.next;
          fast = fast!.next.next;
          if (slow === fast) return true;
        }
        return false;
    }

    //Problem 3 & 4 & 5  - Implementation left for extension.  See comments below.
}


// Add other LinkedList methods here: append, prepend, insertAt, deleteNode etc.
//  Consider adding recursive and iterative solutions for each problem.

//Problem 3: Merge Two Sorted Lists  (Add a mergeSortedLists method)
//Problem 4: Remove Nth Node From End (Add a removeNthFromEnd method)
//Problem 5: Palindrome Linked List (Add a isPalindrome method)  Consider using a stack or recursion.


export { LinkedList, Node };
```

File: leetcode-linked-lists-solutions-optimized/src/LinkedList.java
```java
package src;

class Node {
    int data;
    Node next;

    Node(int d) {
        data = d;
        next = null;
    }
}

public class LinkedList {

    Node head;

    // ... (Methods for all 5 problems will go here. See examples below) ...

    // Example: Reverse a Linked List
    public Node reverseList(Node head) {
        Node prev = null;
        Node curr = head;
        Node next = null;
        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }


    // Example: Detect Cycle
    public boolean hasCycle(Node head) {
        Node slow = head;
        Node fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) return true;
        }
        return false;
    }

    // ... (Rest of the methods) ...

}
```